src/core_editor/editor.rs

```diff
@@ -125,20 +147,66 @@ impl Editor {
         self.line_buffer.is_empty()
     }
 
-    pub fn undo(&mut self) -> Option<()> {
-        self.line_buffer.undo()
+    pub fn reset_olds(&mut self) {
+        self.edits = vec![LineBuffer::new()];
+        // NOTE: Leaving the old logic to help with the review
+        // self.old_lines = vec![vec![String::new()]];
+        // self.old_insertion_point = vec![InsertionPoint::new()];
+        self.index_undo = 2;
+    }
+
+    fn get_index_undo(&self) -> usize {
+        if let Some(c) = self.edits.len().checked_sub(self.index_undo) {
+            c
+        } else {
+            0
+        }
     }
 
     pub fn redo(&mut self) -> Option<()> {
-        self.line_buffer.redo()
+        if self.index_undo > 2 {
+            self.index_undo = self.index_undo.checked_sub(2)?;
+            self.undo()
+        } else {
+            None
+        }
     }
 
-    pub fn reset_olds(&mut self) {
-        self.line_buffer.reset_olds()
+    pub fn undo(&mut self) -> Option<()> {
+        // NOTE: Leaving the old logic to help with the review
+        // self.lines = self.old_lines.get(self.get_index_undo())?.clone();
+        // let insertion_point = *self.old_insertion_point.get(self.get_index_undo())?;
+        // self.set_insertion_point(insertion_point.line, insertion_point.offset);
+
+        self.line_buffer = self.edits.get(self.get_index_undo())?.clone();
+
+        if self.index_undo <= self.edits.len() {
+            self.index_undo = self.index_undo.checked_add(1)?;
+        }
+        Some(())
     }
 
     pub fn set_previous_lines(&mut self, is_after_action: bool) -> Option<()> {
-        self.line_buffer.set_previous_lines(is_after_action)
+        self.reset_index_undo();
+
+        // NOTE: Not sure what we are trying to do here. @mazarano help?
+        if self.old_lines.len() > 1
+            && self.old_lines.last()?.concat().trim().split(' ').count()
+                == self.lines.concat().trim().split(' ').count()
+            && !is_after_action
```

Comment:
@mzanrosso Can you help me a bit here. Like can you explain what this check is for? I havent' really applied any brains in this PR. I just did one-to-one mapping of what was already present and what I came up with.

Reply:
@sherubthakur Imagine that you write "te", "tes" and finally "test"; basically we don't want to have 3 lines in the old_lines state vector but we want to update only the last word since we decided to have history by words and not by single character.

Reply:
Ok, so here is my take. if we introduce the following 
- `word count` in `LineBuffer` which gives the count of words in that buffer
- `last_buffer` which gives the last element of the edit buffer
- `current_buffer` which gives handle to the current line buffer
then this will look something like

```rust
if self.has_undo_history() && self.last_buffer()?.word_count() == self.current_buffer().word_count()
```

This would be equivalent to what you intended. Right?

Note: Thoughts on trying to pull the strategy of having swappable logic for how we decide to do history. e.g. word-based, line-based, character-based, etc. Though I am not sure how this will look like.

Reply:
@sherubthakur You miss the last condition `&& !is_after_action` but in theory should be fine